src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -630,28 +630,27 @@ private void CreateGrainInstance(string grainTypeName, ActivationData data, stri
             grain.Data = data;
 
             Type stateObjectType = grainTypeData.StateObjectType;
-            GrainState state;
+            object state = null;
             if (stateObjectType != null)
             {
-                state = (GrainState) Activator.CreateInstance(stateObjectType);
-                state.InitState(null);
-            }
-            else
-            {
-                state = null;
+                state = Activator.CreateInstance(stateObjectType);
             }
 
-            lock (data)
-            {
-                grain.Identity = data.Identity;
-                data.SetGrainInstance(grain);
+            grain.Identity = data.Identity;
+            data.SetGrainInstance(grain);
 
-                if (state != null)
+            var statefulGrain = grain as IStatefulGrain;
```

Comment:
why are we changing locking here?
How is it related to this change? Lets please maintain EXACTLY the same locking. If you feel like locking is wrong/can be improved here, lets do it as a separate PR.